File: Java/Tema6_Ejercicio3.java
/*
 * TEMA 6: CADENAS DE CARACTERES - EJERCICIO 3
 * ============================================
 * Escriba un programa que busque una subcadena dentro de una cadena.
 */

import java.util.Scanner;

public class Tema6_Ejercicio3 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Digite la cadena principal: ");
        String cadena = scanner.nextLine();
        
        System.out.print("Digite la subcadena a buscar: ");
        String subcadena = scanner.nextLine();
        
        System.out.println("\n=== BÚSQUEDA DE SUBCADENA ===");
        System.out.println("Cadena: '" + cadena + "'");
        System.out.println("Buscar: '" + subcadena + "'");
        
        if (cadena.contains(subcadena)) {
            int indice = cadena.indexOf(subcadena);
            System.out.println("Encontrado en la posición: " + indice);
            System.out.println("Conteo de ocurrencias: " + countOccurrences(cadena, subcadena));
        } else {
            System.out.println("No encontrado.");
        }
        
        scanner.close();
    }
    
    static int countOccurrences(String text, String pattern) {
        int count = 0;
        int index = 0;
        while ((index = text.indexOf(pattern, index)) != -1) {
            count++;
            index += pattern.length();
        }
        return count;
    }
}

/*
 * SALIDA ESPERADA:
 * Digite la cadena principal: banana
 * Digite la subcadena a buscar: ana
 * 
 * === BÚSQUEDA DE SUBCADENA ===
 * Cadena: 'banana'
 * Buscar: 'ana'
 * Encontrado en la posición: 1
 * Conteo de ocurrencias: 2
 */

File: JS/Tema6_Ejercicio3.js
/*
 * TEMA 6: CADENAS DE CARACTERES - EJERCICIO 3
 * ============================================
 * Contar y buscar subcadenas.
 * 
 * EXPLICACIÓN:
 * - indexOf(): busca una cadena
 * - lastIndexOf(): última ocurrencia
 * - Contar repeticiones con bucle
 */

function contarOcurrencias(texto, subcadena) {
    let contador = 0;
    let pos = 0;
    
    while ((pos = texto.indexOf(subcadena, pos)) !== -1) {
        contador++;
        pos += subcadena.length;
    }
    
    return contador;
}

const texto = "JavaScript es genial. Aprender JavaScript es muy rewarding.";
const palabra = "JavaScript";

console.log("=== BÚSQUEDA Y CONTEO ===");
console.log(`Texto: "${texto}"`);

const ocurrencias = contarOcurrencias(texto, palabra);
console.log(`\nOcurrencias de '${palabra}': ${ocurrencias}`);

console.log(`Primera ocurrencia: posición ${texto.indexOf(palabra)}`);
console.log(`Última ocurrencia: posición ${texto.lastIndexOf(palabra)}`);

// Buscar múltiples palabras
const palabras = ['es', 'JavaScript', 'learning'];
console.log(`\nBúsqueda múltiple:`);
palabras.forEach(p => {
    const count = contarOcurrencias(texto, p);
    console.log(`- '${p}': ${count} veces`);
});

/*
 * SALIDA ESPERADA:
 * === BÚSQUEDA Y CONTEO ===
 * Texto: "JavaScript es genial. Aprender JavaScript es muy rewarding."
 * 
 * Ocurrencias de 'JavaScript': 2
 * Primera ocurrencia: posición 0
 * Última ocurrencia: posición 33
 * 
 * Búsqueda múltiple:
 * - 'es': 2 veces
 * - 'JavaScript': 2 veces
 * - 'learning': 0 veces
 */
